File: src/middlewares/errorHandler.js
const logger = require('../utils/logger');

const errorHandler = (err, req, res, next) => {
  logger.error('Error occurred:', {
    message: err.message,
    stack: process.env.NODE_ENV === 'development' ? err.stack : undefined,
    path: req.path,
    method: req.method
  });

  const statusCode = err.statusCode || 500;
  const message = err.message || 'Internal server error';
  const code = err.code || 'INTERNAL_ERROR';

  const response = {
    error: {
      message,
      code
    }
  };

  if (err.details && Array.isArray(err.details)) {
    response.error.details = err.details;
  }

  if (process.env.NODE_ENV === 'development' && err.stack) {
    response.error.stack = err.stack;
  }

  res.status(statusCode).json(response);
};

module.exports = errorHandler;

File: file-storage-service/internal/handler/middleware/error_handler.go
package middleware

import (
	"file-storage-service/pkg/response"

	"github.com/gofiber/fiber/v2"
	"go.uber.org/zap"
)

func ErrorHandler(log *zap.Logger) fiber.ErrorHandler {
	return func(c *fiber.Ctx, err error) error {
		code := fiber.StatusInternalServerError
		if e, ok := err.(*fiber.Error); ok {
			code = e.Code
		}

		log.Error("unhandled error",
			zap.Error(err),
			zap.String("method", c.Method()),
			zap.String("path", c.Path()),
		)

		return response.Err(c, code, "INTERNAL_ERROR", err.Error())
	}
}
